### PR TITLE
constructable lightswitches

### PR DESCRIPTION
**Pull request by Skeletonman0**
[OBJECTS] [QOL]
## About the PR 
makes light switches mechscannable, and moves their auto pixel shifting to its own proc.
also some small tweaks to make it work consistently

## Why's this needed? 
buildable stuff good, was slightly annoying during some of the crash map rounds that you couldn't get new ones

### CODE REVIEW

File: code/obj/machinery/lightswitch.dm [hunk 1]
```diff
@@ -49,6 +50,29 @@
 	if(!ON_COOLDOWN(src, "mechcomp_toggle", 1 SECOND))
 		toggle(null)
 
+/obj/machinery/light_switch/proc/autoposition()
+	var/turf/T = null
+	SPAWN(1 DECI SECOND)
```
Comment (BatElite) [inline code/obj/machinery/lightswitch.dm:55]: It might be slightly neater code to keep this spawn in `/obj/machinery/light_switch/auto/New()`, built switches don't need to delay for walls spawning in.

File: code/obj/machinery/lightswitch.dm [hunk 2]
```diff
@@ -14,16 +14,17 @@
...
 
 /obj/machinery/light_switch/New()
 	..()
 	UnsubscribeProcess()
-	light = new /datum/light/point
+	if (!light)
+		light = new /datum/light/point
 	SPAWN(0.5 SECONDS)
-		src.area = src.loc.loc
+		src.area = get_area(src)
```
Comment (BatElite) [inline code/obj/machinery/lightswitch.dm:27]: Not that you added that spawn, but I think this means that there's a half second window for runtimes to happen because area isn't set when toggling (that's normally covered up by the pregame) or toggling the previously assigned area.

I'm kinda feeling like calling New() in was_built_from_frame isn't great in general (kinda inelegant for deconned and re-built switches, a double call for completely new switches), but I'm not sure if that's actually an issue or if it's just me going "I personally wouldn't code it like that so it sucks".

Comment (TobleroneSwordfish) [reply code/obj/machinery/lightswitch.dm:27]: Yeeah `src.New()` gives me bad vibes

File: code/obj/machinery/lightswitch.dm [hunk 3]
```diff
@@ -49,6 +50,29 @@
...
+				if (dir == EAST)
+					src.pixel_x = 24
+				else if (dir == WEST)
+					src.pixel_x = -24
+				else if (dir == NORTH)
+					src.pixel_y = 24
+				else
+					src.pixel_y = -24
+				break
+		T = null
```
Comment (TobleroneSwordfish) [inline code/obj/machinery/lightswitch.dm:69]:
```suggestion
```

**Comment by Skeletonman0** [comment]
> Does this allow you to decon them too?

should only allow you to decon built ones

File: code/obj/machinery/lightswitch.dm [hunk 4]
```diff
@@ -49,6 +50,36 @@
 	if(!ON_COOLDOWN(src, "mechcomp_toggle", 1 SECOND))
 		toggle(null)
 
+/obj/machinery/light_switch/proc/autoposition()
+	var/turf/T = null
+	for (var/dir in cardinal)
+		T = get_step(src,dir)
+		if (istype(T,/turf/simulated/wall) || (locate(/obj/wingrille_spawn) in T) || (locate(/obj/window) in T))
```
Comment (pali6) [inline code/obj/machinery/lightswitch.dm:57]: Imo check T.density instead of subtype.

File: code/obj/machinery/lightswitch.dm [hunk 5]
```diff
@@ -14,16 +14,17 @@
...
-	// mats = 6 fuck you mport
+	mats = list("MET-1"=10,"CON-1"=15)
 	var/datum/light/light
 
 
 /obj/machinery/light_switch/New()
 	..()
 	UnsubscribeProcess()
-	light = new /datum/light/point
+	if (!light)
```
Comment (pali6) [inline code/obj/machinery/lightswitch.dm:24]: Why is this necessary?

File: code/obj/machinery/lightswitch.dm (hunk 4 above)
Comment (pali6) [inline code/obj/machinery/lightswitch.dm:57]:
```suggestion
		if (T.density || (locate(/obj/wingrille_spawn) in T) || (locate(/obj/window) in T))
```

**Review by pali6 — APPROVED** [review]
